### PR TITLE
Class2

### PR DESCRIPTION
**Pull request by AlexLoitzl**

### CODE REVIEW

File: class_2.tex [hunk 1]
```diff
@@ -106,55 +119,88 @@ \subsection{Lattices and Fixpoints}
...
+    Let $(A,\sqsubseteq)$ be a poset. Then (i) $x$ in $A$ is an \emph{upper bound} of a subset $B$ of $A$ if for all $y$ in $B$, it holds that $y \sqsubseteq x$;
+    (ii) $x$ is the \emph{least upper bound} (lub) of $B$ if $x$ is an upper bound of $B$ and for all upper bounds $y$ of $B$, we have $x \sqsubseteq y$. %We denote such $x$ by $\bigsqcup B$.
 \end{definition}
 
 \begin{definition}[Lower Bound]
-    Let $(A,\sqsubseteq)$ be a poset. Then (i) $x$ in $A$ is an \emph{lower bound} on a subset $B$ of $A$ if for all $y$ in $B$, it holds that $y \sqsubseteq x$; $x$ is the greatest lower bound of $B$ if $x$ is a lower bound of $B$ and for all lower bounds $y$ of $B$, we have $y \sqsubseteq x$. We denote such $x$ by $\bigsqcap B$.
+    Let $(A,\sqsubseteq)$ be a poset. Then (i) $x$ in $A$ is a \emph{lower bound} of a subset $B$ of $A$ if for all $y$ in $B$, it holds that $x \sqsubseteq y$; $x$ is the \emph{greatest lower bound} (glb) of $B$ if $x$ is a lower bound of $B$ and for all lower bounds $y$ of $B$, we have $y \sqsubseteq x$. %We denote such $x$ by $\bigsqcap B$.
```
Comment (smml1996) [inline class_2.tex]: here can we add point ii) ? (like you did for upper bound)

File: class_2.tex [hunk 2]
```diff
@@ -106,55 +119,88 @@ \subsection{Lattices and Fixpoints}
...
 \end{definition}
 
-\begin{example} 
-    Consider the poset $(\NN,\leq)$. Then for any $B \subseteq \NN$, if $B$ is finite, $\bigsqcup B$ is well-defined and equal to $\max B$. If $B$ is infinite, then $\bigsqcup B$ does not exist. 
+\begin{example}
+    Consider the poset $(\NN,\leq)$. Then for any $B \subseteq \NN$, if $B$ is finite, $\lub(B)$ is well-defined and equal to $\max B$. If $B$ is infinite, then $\lub(B)$ does not exist.
```
Comment (smml1996) [inline class_2.tex]: here are we missing some parenthesis like \max (B)? (if it was on purpose never mind)

File: class_2.tex [hunk 3]
```diff
@@ -106,55 +119,88 @@ \subsection{Lattices and Fixpoints}
...
+        y \sqsubseteq z$.
+    \end{enumerate}
+    From (ii) we know that $\hat{y}$ is an upper bound, and therefore, by (iii), $\hat{x} \sqsubseteq \hat{y}$. Symmetrically, by (i) and (iv) we get $\hat{y} \sqsubseteq \hat{x}$. By antisymmetry, $x = y$.
+\end{proof}
+
+Given that the \emph{least upper bound} and the \emph{greatest lower bound} are unique (if they exist), for any set $A$ and $B \subseteq A$, we write $\bigsqcup B$ to denote $\lub(B)$ and $\bigsqcap B$ to denote $\glb(B)$.
+
+\begin{definition}[Lattice]
+    A poset $(A,\sqsubseteq)$ is a \emph{lattice} if for all $x$, $y$ in $A$, $\bigsqcup \{x,y\}$ ($x \sqcup y$, join) and $\bigsqcap \{x,y\}$ ($x \sqcap y$, meet) exist.
+\end{definition}
```
Comment (smml1996) [inline class_2.tex:166]: maybe we can add some plain english between the comma in .`...$A$, $\bigsqcup ..`

File: class_2.tex [hunk 4]
```diff
@@ -106,55 +119,88 @@ \subsection{Lattices and Fixpoints}
...
 
 \begin{definition}[Complete Lattice]
-    Poset $(A,\sqsubseteq)$ is a \emph{complete-lattice} if for all $B \subseteq A$, both $\bigsqcap B$ and $\bigsqcup B$ exist.
+    A poset $(A,\sqsubseteq)$ is a \emph{complete lattice} if for all $B \subseteq A$, $\bigsqcap B$ and $\bigsqcup B$ exist.
```
Comment (smml1996) [inline class_2.tex]: similar here

File: class_2.tex [hunk 5]
```diff
@@ -106,55 +119,88 @@ \subsection{Lattices and Fixpoints}
...
+    Let $(A,\sqsubseteq)$ be a complete lattice and let $F$ a monotone function on $A$.
+    Further, let $\hat{y} = \{x \mid x \sqsubseteq F(x)\}$ and $\hat{z} = \{x \mid F(x) \sqsubseteq x\}$.
+    It holds, that:
+    \begin{enumerate}
+      \item $\hat{y}$ and $\hat{z}$ are fixpoints of $F$,
+      \item for all fixpoints $x$ of $F$, $\hat{y} \sqsubseteq x \sqsubseteq \hat{z}$
+    \end{enumerate}
 \end{theorem}
 
+{\flushleft We write $\lfp(F) = \{x \mid x \sqsubseteq F(x)\}$ and $\gfp(F) = \{x \mid F(x)\sqsubseteq x\}$, to denote the least fixpoint and the greatest fixpoint, respectively.}
```
Comment (smml1996) [inline class_2.tex]: here the greatest fixpoint should be the set of the least fixpoint and viceversa. Also you are missing to take the least upper bound and the greatest lower bounds on these sets

**Review by smml1996 — CHANGES_REQUESTED** [review]
Great work! just minor fixes!

File: class_2.tex (hunk 2 above)
Comment (AlexLoitzl) [reply class_2.tex]: I added them for consistency

File: class_2.tex (hunk 5 above)
Comment (AlexLoitzl) [reply class_2.tex]: Thanks... That was stupid

File: class_2.tex [hunk 6]
```diff
@@ -106,55 +119,89 @@ \subsection{Lattices and Fixpoints}
...
+% TODO
+\begin{definition}[Fixpoint]
+Let $(A, \sqsubseteq)$ be a poset and $F: A \to A$. An element $x$ in $A$ is a \emph{fixpoint} of $F$ if it is both a \emph{pre-} and \emph{postfixpoint} of $F$.
+\end{definition}
 
 \begin{theorem}[Knaster-Tarski]
 \label{thrm:knaster}
-    For every complete lattice $(A,\sqsubseteq)$ and monotone function $f$ on $A$, it holds that: (i) $\bigsqcup \{x\in A| x \sqsubseteq f(x)\}$ is the unique greatest fixpoint of $F$; (ii) $\bigsqcap \{x \in A| f(x) \sqsubseteq x\}$ is the unique least fixpoint of $f$. 
+    Let $(A,\sqsubseteq)$ be a complete lattice and let $F$ a monotone function on $A$.
+    Further, let $\hat{y} = \{x \mid x \sqsubseteq F(x)\}$ and $\hat{z} = \{x \mid F(x) \sqsubseteq x\}$.
```
Comment (smml1996) [inline class_2.tex]: here you are saying that \hat{y} and \hat{z} are subsets of A but below you are saying they are fixpoints

File: class_2.tex [hunk 7]
```diff
@@ -106,55 +119,89 @@ \subsection{Lattices and Fixpoints}
...
+    It holds, that:
+    \begin{enumerate}
+      \item $\hat{y}$ and $\hat{z}$ are fixpoints of $F$,
+      \item for all fixpoints $x$ of $F$, $\hat{y} \sqsubseteq x \sqsubseteq \hat{z}$
```
Comment (smml1996) [inline class_2.tex]: is the relation between z,x,y correct?  🤔

**Review by smml1996 — CHANGES_REQUESTED** [review]

File: class_2.tex (hunk 7 above)
Comment (AlexLoitzl) [reply class_2.tex]: Sorry about that... I think I left too much from the old ones and messed it up. I think it's correct now

**Review by smml1996 — APPROVED** [review]
now looks great!!

**Pull request merged**